packages/http/src/router/matchBaseUrl.ts

```diff
@@ -1,23 +1,45 @@
+import * as E from 'fp-ts/lib/Either';
+import * as O from 'fp-ts/lib/Option';
+import { pipe } from 'fp-ts/lib/pipeable';
+
 import { INodeVariable, IServer } from '@stoplight/types';
 import { MatchType } from './types';
 
 const variableRegexp = /{(.*?)}/g;
 
-export function matchBaseUrl(server: IServer, baseUrl: string): MatchType {
-  const templateMatchResult = matchRequestUrlToTemplateUrl(baseUrl, server.url, server.variables);
+export function matchBaseUrl(server: IServer, baseUrl: string): E.Either<Error, MatchType> {
+  return pipe(
+    convertTemplateToRegExp(server.url, server.variables),
+    E.chain(regex =>
+      pipe(
+        regex.exec(baseUrl),
+        E.fromNullable(new Error('No matches')),
+        E.map(matches => (matches.length > 1 ? MatchType.TEMPLATED : MatchType.CONCRETE)),
+        E.orElse(() => E.right<Error, MatchType>(MatchType.NOMATCH))
+      )
+    )
+  );
+}
 
-  if (!templateMatchResult) {
-    return MatchType.NOMATCH;
-  }
+export function convertTemplateToRegExp(
+  urlTemplate: string,
+  variables?: { [name: string]: INodeVariable }
+): E.Either<Error, RegExp> {
+  return pipe(
+    O.fromNullable(variables),
+    O.fold(
+      () => E.right(urlTemplate),
+      vars => replaceString(vars, urlTemplate)
+    ),
+    E.map(regexString => new RegExp(`^${regexString}$`))
+  );
```

Comment:
Try to stick to the same monad whenever is possible:

```ts
  return pipe(
    variables,
    E.fromNullable(new Error('no variables')),
    E.fold(() => E.right(urlTemplate), vars => replaceString(vars, urlTemplate)),
    E.map(regexString => new RegExp(`^${regexString}$`))
  );
```

However since the `E.fromNullable` here is doing nothing else than discriminating the starting `Either`, we can also write

```ts
  return pipe(
    variables ? replaceString(variables, urlTemplate) : E.right(urlTemplate),
    E.map(regexString => new RegExp(`^${regexString}$`))
  );

```

Reply:
Actually please use the latter version

```ts
  return pipe(
    variables ? replaceString(variables, urlTemplate) : E.right(urlTemplate),
    E.map(regexString => new RegExp(`^${regexString}$`))
  );
```

Reply:
Makes sense, thank you.